### PR TITLE
feat: add secrets / config to a verbs metadata

### PR DESCRIPTION
**Pull request by stuartwdouglas**
fixes: #1083

This just adds them for JVM languages for now, as the go implementation will change when they move to an injection model.

### CODE REVIEW

File: cmd/go2proto/main.go [hunk 1]
```diff
@@ -130,6 +130,7 @@ func genErrorf(pos token.Pos, format string, args ...any) error {
 
 var tmpl = template.Must(template.New("proto").
 	Parse(`
+// THIS FILE IS GENERATED; DO NOT MODIFY
```
Comment (alecthomas) [inline cmd/go2proto/main.go:133]: Good call.

File: internal/schema/metadataconfig.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	Config []*Ref `parser:"'+' 'config' @@ (',' @@)*" protobuf:"2"`
+}
+
+var _ Metadata = (*MetadataConfig)(nil)
+
+func (m *MetadataConfig) Position() Position { return m.Pos }
+func (m *MetadataConfig) String() string {
+	out := &strings.Builder{}
+	fmt.Fprint(out, "+calls ")
```
Comment (alecthomas) [inline internal/schema/metadataconfig.go:26]:
```suggestion
	fmt.Fprint(out, "+config ")
```

File: internal/schema/metadataconfig.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	w := 6
+	for i, call := range m.Config {
```
Comment (alecthomas) [inline internal/schema/metadataconfig.go:28]: s/call/config/g

File: internal/schema/metadatasecrets.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+	Secrets []*Ref `parser:"'+' 'config' @@ (',' @@)*" protobuf:"2"`
+}
+
+var _ Metadata = (*MetadataSecrets)(nil)
+
+func (m *MetadataSecrets) Position() Position { return m.Pos }
+func (m *MetadataSecrets) String() string {
+	out := &strings.Builder{}
+	fmt.Fprint(out, "+calls ")
```
Comment (alecthomas) [inline internal/schema/metadatasecrets.go:26]:
```suggestion
	fmt.Fprint(out, "+secrets ")
```

File: internal/schema/metadatasecrets.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	schemapb "github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/schema"
+)
+
+// MetadataSecrets represents a metadata block with a list of config items that are used.
+//
+//protobuf:11,optional
+type MetadataSecrets struct {
+	Pos Position `parser:"" protobuf:"1,optional"`
+
+	Secrets []*Ref `parser:"'+' 'config' @@ (',' @@)*" protobuf:"2"`
```
Comment (alecthomas) [inline internal/schema/metadatasecrets.go:18]:
```suggestion
	Secrets []*Ref `parser:"'+' 'secrets' @@ (',' @@)*" protobuf:"2"`
```

File: internal/schema/metadatasecrets.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+var _ Metadata = (*MetadataSecrets)(nil)
+
+func (m *MetadataSecrets) Position() Position { return m.Pos }
+func (m *MetadataSecrets) String() string {
+	out := &strings.Builder{}
+	fmt.Fprint(out, "+calls ")
+	w := 6
+	for i, call := range m.Secrets {
```
Comment (alecthomas) [inline internal/schema/metadatasecrets.go:28]: s/call/ref/g

**Review by alecthomas — APPROVED** [review]
This is great, thanks Stuart.

File: internal/schema/metadataconfig.go (hunk 2 above)
Comment (stuartwdouglas) [reply internal/schema/metadataconfig.go]: Ugh, sorry, this PR is mostly just copypasta from from the calls metadata.

File: internal/schema/validate.go [hunk 7]
```diff
@@ -341,8 +341,11 @@ func ValidateModule(module *Module) error {
 
 		case *Data:
 			for _, md := range n.Metadata {
-				if md, ok := md.(*MetadataCalls); ok {
+				switch md.(type) {
+				case *MetadataCalls, *MetadataSecrets, *MetadataConfig:
```
[the remark this answers is not in the document]
Comment (stuartwdouglas) [reply internal/schema/validate.go:345]: Yea, there are definitely gaps in the current logic, I don't think it is high priority though.

File: internal/schema/metadatasecrets.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+	Secrets []*Ref `parser:"'+' 'secrets' @@ (',' @@)*" protobuf:"2"`
+}
+
+var _ Metadata = (*MetadataSecrets)(nil)
+
+func (m *MetadataSecrets) Position() Position { return m.Pos }
+func (m *MetadataSecrets) String() string {
+	out := &strings.Builder{}
+	fmt.Fprint(out, "+secrets ")
```
Comment (alecthomas) [inline internal/schema/metadatasecrets.go:26]: I forgot to say, there are tests that should be updated to exercise this parsing, in particular TestParserRoundtrip et al and TestProtoRoundTrip.

**Pull request merged**